DelayedClipboard/DelayedClipboardRenderingExplainer.md

```diff
@@ -150,11 +154,72 @@ const clipboardItemInput = new ClipboardItem({['image/png']: blobInput1, ['text/
+
+**Problems with this approach**
+Could slow down the browser/tab close operation. The timeout is a way to mitigate this performance issue, but it could lead to other problems like the web author may not have enough time to generate the payload for the format, so the clipboard would have empty data for that format.
+
+3. Choose to only keep the formats that are cheaper to produce and remove the rest (the formats that are delay rendered) by performing another clipboard.write operation: Native Excel has some logic to clear the clipboard and write a smaller subset of the initial formats that were on the clipboard when a user decides to close the app. Web authors can do this as well, but it would affect the experience of the user if they are unable to paste the format that has high fidelity content and is delay rendered by the web author. Also, this has to be done during `beforeunload` event which affects performance of the site and slow down browser/tab close operation.
+
+4. Throw away all the delay rendered formats and put empty data for those formats in the clipboard: It would be a bad user experience, but a web author could also choose to not populate the formats that were registered for delay rendering.
+
+5. Browsers could choose to show a confirmation dialog if there is any delay rendered formats in the clipboard: Although it is similar to option 1, here the text of the dialog explicitly mentions delay rendering of formats which is more informative to the the user. This option has similar concerns as option 1, but it does affect navigations, so it slows down browser/tab shutdown behavior.
+
+More info [here](https://docs.google.com/document/d/1H6ow7RWa4MeycKP3OQBoMhLuMaJct3jSCWW6apcVu9U/edit?usp=sharing)
+
+### Proposal
+
+Proposal is a hybrid of options 2 and 4
+
+1. At least one built-in format must be written to the clipboard without delay as part of the write operation so there will be some data on the clipboard (could be a low fidelity `text/plain` format) if the browser/tab closes and the site doesn't get a chance to generate the data for the delay rendered formats within the timeout window. If the site only writes one format to the clipboard, then it can't be delay rendered.
+
+2. If a delay rendered callback is already running before the page unloads, cancel the callback after a timeout period if the callback hasn't completed yet. When the callback is cancelled, set empty data to the clipboard for that delay rendered format.
+
+3. If site registered for `beforeunload` event to run the callback, run the callback but cancel it if it exceeds a timeout period so it doesn't cause delays in navigation.
+
 ## Privacy and Security Considerations
 
 ### Privacy
```

Comment:
Here's a privacy concern you may or may not have already considered. Sorry if this is already captured somewhere.

Consider, I dunno, a dedicated user of WordPerfect or some other once-popular, now-esoteric application. A malicious web site says that it can generate a WordPerfect document. The user cuts some content from the web site and pastes it into WordPerfect. The browser now calls the WordPerfect format callback, and the website now knows that this user is one of the like 5 remaining humans who use that tool. This seems like some very interesting entropy from a fingerprinting perspective.

Now imagine, instead of WordPerfect, the scenario is instead some domain-specific application which is popular amongst some marginalized communit{y,ies}. Now we're revealing to the site that the user is probably a member of said communit{y,ies}.

Reply:
@hober Thank you for your comment and highlighting the privacy issue! I did capture a somewhat similar issue in the section below with an image that describes this privacy issue in more detail. Please let me know if I'm missing anything here. Thanks! 

There are few things that we want to mention here:
1. During copy, at least in Chromium, the clipboard permission prompt serves as a trust mechanism which means the user trusts the web app to access the system clipboard. If a malicious site tricks the user into obtaining this trust, then it is really hard to prevent that site from gaining access to the system clipboard.
2. When the malicious site tricks the user into copying WordPerfect format, the user *has* to paste it in an app that reads that format, otherwise the format never gets read and the callback never gets triggered.
3. If we restrict the number of web custom formats to 1 or 2, then we limit the apps that can be detected by sites that write web custom formats for fingerprinting purposes. Note that if the source app doesn't produce the right data for the web custom format during paste, then the paste would fail and the user will know that the site doesn't support this format, so it's hard for user to trust this site after the initial copy-paste operation.
4. There could be more than one app that supports the WordPerfect format, so at best it gives some fuzzy details about the app ecosystem where the paste happened.

@evanstade @dandclark @sanketj anything you would like to add here?

Reply:
Also, added the example you mentioned in your comment. We can discuss more in https://github.com/w3c/editing/issues/439. Thanks!